factor/totp/tests/factor_test.php

```diff
@@ -113,4 +117,27 @@ public function test_validate_code() {
         $result = $totpfactor->validate_code($code, $window, $totp, $factorinstance);
         $this->assertEquals($totpfactor::TOTP_INVALID, $result);
     }
+
+    /**
+     * Do not store the TOTP secret + user combination more than once
+     *
+     * @covers ::setup_user_factor
+     */
+    public function test_wont_store_same_secret_twice() {
+        $this->resetAfterTest(true);
+        $user = $this->getDataGenerator()->create_user();
+        $this->setUser($user);
+
+        set_config('enabled', 1, 'factor_totp');
+        $totpfactor = \tool_mfa\plugininfo\factor::get_factor('totp');
+        $totpdata = [
+            'secret' => 'fakekey',
+            'devicename' => 'fakedevice',
+        ];
+        $record = $totpfactor->setup_user_factor((object) $totpdata);
+
+        // Trying to add the same TOTP should return the existing record.
+        $anotherecord = $totpfactor->setup_user_factor((object) $totpdata);
```

Comment:
Here there should also be a DB->count_records = 1 perhaps?

Reply:
Done